### PR TITLE
[tf-mlir-translate] patch for handling data types with subtypes

### PR DESCRIPTION
**Pull request by parsabee**
As mentioned in this issue: https://github.com/tensorflow/tensorflow/issues/50521#issuecomment-871579698, tf-mlir-translate didn't have a mechanism for handling data type with subtypes such as DT_RESOURCE. This patch adds support for handling types that might have subtypes to tf-mlir-translate and it's python wrapper.

### CODE REVIEW

File: tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.h [hunk 1]
```diff
@@ -38,6 +38,12 @@ struct ArrayInfo {
   TensorShapeProto shape;
 };
 
+struct ArrayInfo : public ArrayInfoBase {
```
Comment (Mogball) [inline tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.h:41]: Why did you add a subclass? I don't see the base case used anywhere.

Comment (Mogball) [inline tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.h:41]: Why did you add a subclass? I don't see the base case used anywhere.

File: tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.cc [hunk 2]
```diff
@@ -190,11 +244,54 @@ Status ParseNodeNames(absl::string_view names_str,
...
+  }
+  return dtypes;
+}
+
 Status ParseNodeDataTypes(absl::string_view data_types_str,
                           std::vector<std::string>& data_type_vector) {
   data_type_vector.clear();
   if (!data_types_str.empty()) {
-    data_type_vector = absl::StrSplit(data_types_str, ',');
+    auto s = ParseDTypesHelper(data_types_str);
```
Comment (Mogball) [inline tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.cc:290]: TF_RETURN_IF_ERROR?

Comment (Mogball) [inline tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.cc:290]: TF_ASSIGN_OR_RETURN?

The program shouldn't die if there was a syntax error in the input.

File: tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.cc [hunk 3]
```diff
@@ -190,11 +244,54 @@ Status ParseNodeNames(absl::string_view names_str,
...
+  int cur_pos = 0;
+  std::vector<std::string> dtypes;
+  for (auto it : llvm::enumerate(data_types_str)) {
+    char c = it.value();
+    int i = it.index();
+    // Skip parsing the subtypes of a type
+    if (c == '(') {
+      if (inside_subtype) {
+        return errors::FailedPrecondition(
+            absl::StrCat("Syntax error in data types '", data_types_str, "'"));
```
Comment (Mogball) [inline tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.cc:259]: Can you add more descriptive error messages? At least something like "Unexpected '('".

File: tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.cc [hunk 4]
```diff
@@ -165,20 +228,11 @@ Status ParseNodeShapes(
...
-        }
-        int size;
-        TF_RET_CHECK(absl::SimpleAtoi(dim_str, &size));
-        dims.push_back(size);
+      auto s = ParseShapeStr(node_shapes_str[i]);
+      if (!s.ok()) {
+        return s.status();
       }
-      shapes_vector.push_back(dims);
+      shapes_vector.push_back(s.ValueOrDie());
```
Comment (Mogball) [inline tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.cc:235]: TF_ASSIGN_OR_RETURN

File: tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.cc [hunk 5]
```diff
@@ -190,11 +244,54 @@ Status ParseNodeNames(absl::string_view names_str,
   return Status::OK();
 }
 
+StatusOr<std::vector<std::string>> ParseDTypesHelper(
```
Comment (Mogball) [inline tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.cc:247]: should this be `static`?

File: tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.cc [hunk 6]
```diff
@@ -136,7 +183,23 @@ Status ParseInputArrayInfo(
...
+    std::vector<std::string> parts =
+        absl::StrSplit(type, absl::ByAnyChar("()"));
+    /// If type has subtypes then part[0] = type, part[1] = subtypes, part[2] =
+    /// ""
+    if (parts.size() != 3 && parts.size() != 1) {
+      return errors::InvalidArgument("Invalid type '", type, "'");
+    } else if (parts.size() == 3) {
+      // First part is the type, second is the subtype
+      ArrayInfo::SubTypeInfo subtype;
+      auto s = HandleSubtype(parts[1], &subtype);
```
Comment (Mogball) [inline tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.cc:196]: TF_RETURN_IF_ERROR

Comment (Mogball) [inline tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.cc:196]: TF_RETURN_IF_ERROR

File: tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.cc [hunk 7]
```diff
@@ -92,6 +93,51 @@ Status ParseInputArrayInfo(absl::string_view array_names,
   return ParseInputArrayInfo(node_names, node_dtypes, node_shapes, inputs);
 }
 
+StatusOr<std::vector<int>> ParseShapeStr(absl::string_view node_shapes_str) {
```
Comment (Mogball) [inline tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.cc:96]: should this be `static`?

Comment (Mogball) [inline tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.cc:96]: should this be `static`?

**Review by Mogball — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/mlir/tensorflow/translate/import_model.cc [hunk 8]
```diff
@@ -2599,8 +2599,31 @@ StatusOr<mlir::FunctionType> GraphDefImporter::InferMainFunctionType(
         return errors::InvalidArgument("Input ", i, "has invalid data type");
       }
     }
-    TF_RETURN_IF_ERROR(
-        ::tensorflow::ConvertDataType(imported_dtype, builder, &element_type));
+    /// Check if we have subtypes first
```
Comment (Mogball) [inline tensorflow/compiler/mlir/tensorflow/translate/import_model.cc:2602]: Use two backslashes for comments

File: tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.h (hunk 1 above)
Comment (parsabee) [reply tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.h:41]: So initially I just wanted to just add a 'std::vector<ArrayInfo> subtypes' to the ArrayInfo class, however since the subtypes are not going to have subtypes themselves, I decided to factor out shape and imported_dtype into a base class and alias SubTypeInfo to be the base class, and have ArrayInfo inherit from base and have a std::vector<SubTypeInfo>. This was the cleanest solution I could think of

File: tensorflow/compiler/mlir/tensorflow/translate/import_model.cc [hunk 9]
```diff
@@ -2599,8 +2599,31 @@ StatusOr<mlir::FunctionType> GraphDefImporter::InferMainFunctionType(
...
+    if (!node_info.subtypes.empty()) {
+      std::vector<mlir::TensorType> subtypes;
+      for (const auto& st : node_info.subtypes) {
+        mlir::Type t;
+        llvm::SmallVector<int64_t, 4> shape;
```
Comment (Mogball) [inline tensorflow/compiler/mlir/tensorflow/translate/import_model.cc:2607]: Drop the explicit small vector shape unless you have a good reason to manually specify 4

File: tensorflow/compiler/mlir/tensorflow/translate/import_model.cc [hunk 10]
```diff
@@ -2599,8 +2599,31 @@ StatusOr<mlir::FunctionType> GraphDefImporter::InferMainFunctionType(
...
+        TF_RETURN_IF_ERROR(ConvertToMlirShape(st.shape, &shape));
+        TF_RETURN_IF_ERROR(
+            ::tensorflow::ConvertDataType(st.imported_dtype, builder, &t));
```
Comment (Mogball) [inline tensorflow/compiler/mlir/tensorflow/translate/import_model.cc:2610]: Why do you need `::tensorflow::`?

File: tensorflow/compiler/mlir/tensorflow/translate/import_model.cc [hunk 11]
```diff
@@ -2599,8 +2599,31 @@ StatusOr<mlir::FunctionType> GraphDefImporter::InferMainFunctionType(
         return errors::InvalidArgument("Input ", i, "has invalid data type");
       }
     }
-    TF_RETURN_IF_ERROR(
-        ::tensorflow::ConvertDataType(imported_dtype, builder, &element_type));
+    /// Check if we have subtypes first
+    if (!node_info.subtypes.empty()) {
+      std::vector<mlir::TensorType> subtypes;
+      for (const auto& st : node_info.subtypes) {
+        mlir::Type t;
```
Comment (Mogball) [inline tensorflow/compiler/mlir/tensorflow/translate/import_model.cc:2606]: Please use a more descriptive variable name. `dtype` or `data_type`?

File: tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.cc [hunk 12]
```diff
@@ -92,6 +93,51 @@ Status ParseInputArrayInfo(absl::string_view array_names,
   return ParseInputArrayInfo(node_names, node_dtypes, node_shapes, inputs);
 }
 
+static StatusOr<std::vector<int>> ParseShapeStr(absl::string_view node_shapes_str) {
+  std::vector<int> dims;
+  for (const absl::string_view dim_str : absl::StrSplit(node_shapes_str, ',')) {
```
Comment (Mogball) [inline tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.cc:98]: drop `const` from `string_view`

File: tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.h (hunk 1 above)
Comment (Mogball) [reply tensorflow/compiler/mlir/tensorflow/translate/mlir_roundtrip_flags.h:41]: Gotcha.

**Review by Mogball — CHANGES_REQUESTED** [review]
After these changes, LGTM

**Review by Mogball — APPROVED** [review]
LGTM. Thanks for the patch!

**Review by Mogball — APPROVED** [review]
